packages/sdk/src/roles/user.ts

```diff
@@ -203,4 +203,89 @@ export class User {
     );
     return await babyzk.genProof(proofGenGadgets.wasm, proofGenGadgets.zkey, input);
   }
+
+  /**
+   * genBabyzkProofWithQuery generates a zk proof for the given credential and proof
+   * query string, using identity commitment stored in this user.
+   */
+  public async genBabyzkProofWithQuery(
+    identityCommitment: bigint,
+    cred: Credential,
+    proofGenGadgets: {
+      wasm: Uint8Array;
+      zkey: Uint8Array;
+    },
+    queryStr: string
+  ): Promise<WholeProof> {
+    const userSecrets = this.getIdentitySliceByIdc(identityCommitment);
+    if (!userSecrets) {
+      throw new Error("No identity slice found");
+    }
+
+    const queryObj = query.parse(queryStr);
+    if (!queryObj.ok) {
+      throw queryObj.error;
+    }
+    const statements: statement.Statement[] = [];
+    for (const expr of queryObj.value.conditions) {
+      const claim = cred.body.tp.claims.find(c => c.name === expr.identifier);
+      if (!claim) {
+        throw new Error("Claim not found in the credential body");
+      }
+      switch (claim.type.tp) {
+        case ClaimTypeEnum.Scalar: {
+          statements.push(
+            new statement.ScalarStatement(
+              claim.type as ScalarType,
+              BigInt((expr.value as query.Range).from),
+              BigInt((expr.value as query.Range).to)
+            )
+          );
+          break;
+        }
+        case ClaimTypeEnum.Property: {
+          const ct = claim.type as PropType;
+          const qs = expr.value as query.Set;
+          if (ct.nEqualChecks !== qs.length) {
+            throw new Error("Property width mismatches with the query set length");
+          }
+
+          statements.push(new statement.PropStatement(ct, qs));
+          break;
+        }
+        case ClaimTypeEnum.Boolean: {
+          let reveal = false;
+          if (expr.operation === "REVEAL") {
+            reveal = true;
+          } else if (expr.operation === "HIDE") {
+            reveal = false;
+          } else {
+            throw new Error("Unsupported operation");
+          }
+          statements.push(new statement.BoolStatement(claim.type as BoolType, reveal));
+          break;
+        }
+        default:
+          throw new Error("Unsupported claim type");
+      }
+    }
+
+    // Ensure all claims have corresponding statements.
+    if (cred.body.tp.claims.length !== statements.length) {
+      throw new Error("Claims and statements mismatch");
+    }
```

Comment:
not sound? Must check 1-to-1 mapping. One can have a query looks like this:

"A in [1,2];A in [1,2];A in [1,2];A in [1,2];A in [1,2];A in [1,2];"

Reply:
this is checked in query.ts. But for clarity, updated validations, and updated error messages and tests